API for the license submittal feature

Ugtan opened the pull request:
API for the existing license submittal feature.

Ugtan commented:
@techytushar @goneall I have added the API for the license submittal feature. It would be really great if you guys could try and give your feedback. Although, the **first** ever time implementing is a lot of steps(i know) but only for the very first time registering.

1. Firstly, I use firefox as my browser if you have firefox as well then just download the geckodriver and put it into PATH. And if you use chrome then change the line 536 in views.py to `browser = webdriver.Chrome()` and then download the chrome driver for your browser from [here](http://chromedriver.chromium.org/downloads). finally, set it in PATH variable using `export PATH=$PATH:$PWD/geckodriver`.
2. make sure to install the requirements again(possibly in a venv) using `pip install -r requirements.txt`
3. create a superuser using `python manage.py createsuperuser` then runserver and go to admin then, create a new application by going to the `Add application` section. Copy the client id and client secret of the app and paste it in secret.py file under getOauthToolKitAppID and secret and fill the other details of the app as follows: 
`User: <admin you created>`
`client type: confidential`
`authorization grant type: resource owner password based`
and SAVE the app.
4. While the server is still running just paste the below command in the new tab of the terminal and it will open your browser and ask you for your github credentials inorder to authorize the app. It then generates a authentication code which is used to generate a access token for github to access github api.
`curl -X POST http://localhost:8000/api/submit_license/ -F 'fullname=<your-fullname>' -F 'shortIdentifier=<your-identifier>' -F 'licenseAuthorName=<license-author>' -F 'userEmail=<your-email>' -F 'text=<text>' -F 'osiApproved=<osi>' -F 'sourceUrl=<url>'`

goneall commented:
@Ugtan If someone wanted to use the API from their server based application, would they need to interactively enter credentials in a web browser?

Ugtan commented:
@goneall This was one of my concerns and I have spent a lot of time finding a solution for the same. The only reason for doing the same is to ensure the user doesn't have to hardcode their credentials along with the API request and it is a bit safer if I could say this as GitHub is involved. Let me know your thoughts. Also, just to mention I have followed [this](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/) tutorial for the whole OAuth process.

techytushar commented:
Its showing me this error, when I log into my GitHub account on the webdriver, do we have to add something to the `redirect_uris` while creating the django application? When I check the `Access Tokens` section I see tokens generated there.

![Screenshot from 2019-06-06 01-50-43](https://user-images.githubusercontent.com/26834658/58987642-8affb400-87fd-11e9-9c81-e9a0f432ed9c.png)

Ugtan commented:
> Its showing me this error, when I log into my GitHub account on the webdriver

The state parameter is actually not required it is just an unguessable random string used to protect against cross-site request forgery attacks as mentioned [here](https://developer.github.com/apps/building-oauth-apps/authorizing-oauth-apps/). And it shouldn't affect the working on the working of the API.

> You have commented a line in the `views.py` file, which defines state, is that creating the problem?

Yes, it wasn't required but I will still include it and will update the branch tomorrow. Also, did you get the response like " the license was submitted successfully" after entering the curl command? @techytushar

techytushar commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+                # This is present only when executing submit license via tests
+                urlType = request.POST["urlType"]
+            statusCode = createIssue(licenseAuthorName, licenseName, licenseIdentifier, licenseComments, licenseSourceUrls, licenseHeader, licenseOsi, token, urlType)
+            returnstatus = str(statusCode)
```
Same here, we can reuse the `statusCode` variable instead of making the `returnstatus`

Ugtan replied:
Agreed.

techytushar commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+
+
+def createIssue(licenseAuthorName, licenseName, licenseIdentifier, licenseComments, licenseSourceUrls, licenseHeader, licenseOsi, token, urlType):
+    """ View for creating an GitbHub issue
```
typo `GitHub`

techytushar commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+    the_match = bool(re.compile(r'^(?!the|The.*$).*$').match(licenseName))
+
+    if not no_comma_match:
+        return 'No commas in the fullname of license or exception.'
```
The message can be changed to `No commas allowed in the ...`, this would make the message more clear.

Ugtan replied:
This was the same message as mentioned in the SPDX website about the description of the fields. Although, I agree it makes the message more clear to understand.

techytushar commented:
Nice work with the code and documentation @Ugtan , I have left a few comments on some minor issues that I found out, I would like your feedback on them. Also yes I was able to submit a license using the API and received the message `Success! The license request has been successfully submitted.` Before I was closing the webdriver window as soon as that error appeared but this time I waited for a while and it made the license submission.

@goneall The licenses submitted using the API, should they also appear on the `app/license_requests/` page ?

Ugtan commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+                result=result
+            )
+            result = "Failed to submit the license."
+            httpstatus = 201
```
Agreed. Good point and I'll update it soon.

Ugtan commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+                xml=xml,
+                result=result
+            )
+            result = "Failed to submit the license."
```
The value of the `result` will be retained as `failed to submit` in case the `statusCode` is not `201`

Ugtan commented:
> Nice work with the code and documentation

Thank you so much :) @techytushar 

> Before I was closing the webdriver window as soon as that error appeared but this time I waited for a while and it made the license submission.

I will definitely improve it.

> The licenses submitted using the API, should they also appear on the `app/license_requests/` page ?

I think it should IMO. Good point tho.

Ugtan commented on src/api/views.py:
```diff
@@ -496,4 +513,161 @@ def check_license(request):
+        githubClientSecret = settings.SOCIAL_AUTH_GITHUB_SECRET
+        #state = ''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(16))
+        url = 'https://github.com/login/oauth/authorize/?client_id={}'.format(githubClientId)
+        browser = webdriver.Firefox()
```
> * The application that is calling the API generates the code and passes that in as a parameter in the POST

This was the first ever approach I tried but discarded it as I thought the user will have to pass the authentication code manually. Nevertheless I think this approach would suit the best.

> To make it easy to test, we could move the webdriver code to a separate method or class to allow an interactive test of the license submittal API. My preference would be a separate class to keep the dependencies in this class to a minimum

If we are asking the user for the authentication code then I don't see any need for using the selenium webdriver which will also make the process of testing way easier.

@goneall I liked your approach, just let me know if you want me to implement it and I will do it right now and update the branch.

goneall replied:
@Ugtan Please go ahead and update it with this approach.  The user can always implement a browser approach to gather the username/password as part of the protocol.  We just need to document this well so that the caller knows what the protocol is.

Ugtan commented:
Hello, I have updated the branch with the requested changes as well as added some tests for the license submittal feature.
The API now uses the approach as suggested by @goneall. So, the user will have to go to `http://github.com/login/oauth/authorize/?client_id=<github-client-id>` copy the authentication code from the URL query string into either the cURL command or into the secret.py file(for tests) under `getAuthCode`. I have also updated some of the dependencies so just make sure you run the `pip install -r requirements.txt` again. I have also run the tests locally, you can see the SC. Also, as suggested by @techytushar, license submitted by API will appear on the license requests from now. Let me know you guys feedback on the improved version of the API. It took me a lot of time then I have anticipated but I really hope you guys like the final product.
![photo6334657388564555983](https://user-images.githubusercontent.com/32966387/59164862-87876800-8b30-11e9-96a6-921885f26db9.jpg)

goneall commented:
The code changes look good to me.

> So, the user will have to go to http://github.com/login/oauth/authorize/?client_id=<github-client-id> copy the authentication code ...

We should add this to the documentation.  There's an API section at the end of the README file.

goneall commented:
I ran into some problems migrating the DB.  After updating the requirements, I ran makmigrations and migrate.  I got the following warning:

```
WARNINGS:
?: (rest_framework.W001) You have specified a default PAGE_SIZE pagination rest_framework setting,without specifying also a DEFAULT_PAGINATION_CLASS.
        HINT: The default for DEFAULT_PAGINATION_CLASS is None. In previous versions this was PageNumberPagination. If you wish to define PAGE_SIZE globally whilst defining pagination_class on a per-view basis you may silence this check.
Operations to perform:
  Apply all migrations: admin, api, app, auth, contenttypes, oauth2_provider, sessions, social_django
Running migrations:
  Applying api.0003_auto_20190609_1535... OK
  Applying app.0008_auto_20190609_1535... OK
  Applying oauth2_provider.0005_auto_20170514_1141... OK
```
When running the test, I got the following:

```
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\commands\test.py", line 29, in run_from_argv
    super(Command, self).run_from_argv(argv)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\commands\test.py", line 62, in handle
    failures = test_runner.run_tests(test_labels)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\test\runner.py", line 601, in run_tests
    old_config = self.setup_databases()
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\test\runner.py", line 546, in setup_databases
    self.parallel, **kwargs
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\test\utils.py", line 187, in setup_databases
    serialize=connection.settings_dict.get('TEST', {}).get('SERIALIZE', True),
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\base\creation.py", line 69, in create_test_db
    run_syncdb=True,
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\core\management\commands\migrate.py", line 204, in handle
    fake_initial=fake_initial,
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\migrations\executor.py", line 115, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\migrations\executor.py", line 145, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\migrations\executor.py", line 244, in apply_migration
    state = migration.apply(state, schema_editor)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\migrations\migration.py", line 129, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\migrations\operations\models.py", line 97, in database_forwards
    schema_editor.create_model(model)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\base\schema.py", line 319, in create_model
    self.execute(sql, params or None)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\base\schema.py", line 136, in execute
    cursor.execute(sql, params)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\utils.py", line 62, in execute
    return self.cursor.execute(sql)
  File "C:\Users\gary\git\spdx-online-tools\venv\lib\site-packages\django\db\backends\sqlite3\base.py", line 326, in execute
    return Database.Cursor.execute(self, query)
django.db.utils.OperationalError: table "app_licenserequest" already exists
```

Ugtan commented:
> I ran into some problems migrating the DB. After updating the requirements, I ran makmigrations and migrate. I got the following warning:

I don't think that warning has anything to do with the error you are getting. The warning is generated due to changes in the django rest framework (3.6.1 ~> 3.9)

Also, I am not able to replicate the error you are getting. Just curious are you using `--keepdb` flag while running the tests? cuz whenever we run the tests django creates a blank database which has no scope of a table being exists already or else there may be an error in the migration file.

If you want to run the submit license API tests just try running `python manage.py test api.tests.SubmitLicenseModelsTests` and if you still get the same error just lemme know. @goneall

goneall commented:
> Just curious are you using --keepdb

No - just running test.  It is probably a problem with my local environment and not a problem with the recent code changes.  I see this is passing the Travis build.  I'll try to debug it later this week.

Ugtan commented:
> It is probably a problem with my local environment

Probably cuz it is working for me, Although we can wait for @techytushar to confirm. Also, just let me know if you face any of the issues while debugging. For now I think I should move on to the next task. 

> We should add this to the documentation. There's an API section at the end of the README file.

Also, I will update the documentation soon. @goneall

techytushar commented:
Yes, the tests are running fine for me and also I am able to submit the license using the API. Nice work so far!!

Ugtan commented:
@techytushar @goneall I just wanted to make sure everything is fine and working. Thanks for the update. 

> The warning that is being showed up is due to the changes in the new version of DRF, IMO this should be an easy fix.

Yeah. I will make a commit to fix it as soon as we are done with this.

> Nice work so far!!

Thank you so much 😊

Ugtan commented:
@goneall @techytushar I have added the documentation for the API as well. Please take a look at it. Thanks

goneall commented:
@Ugtan Looks good - nice work.

@techytushar Just confirm it looks good to you and I'll merge.

techytushar commented on README.md:
```diff
@@ -69,12 +69,9 @@ python src/manage.py test
 ## GitHub Developer Sensitive Data
 
-The `settings.py` file uses sensitive data to work with the GitHub API. For that reason, sensitive data is not checked into source. Due to that lack of data, the following error could rise when running the app:
+The `settings.py` file uses sensitive data to work with the GitHub API. For that reason, sensitive data is not checked into source. Due to that lack of data, the License submittal feature and its API won't be able to run as they require the user credentials in order to access the GitHub API. So, the user is supposed to enter their credentials in the `src/src/secret.py ` file in order to ensure pr ...
```
> Due to that lack of data, the License submittal feature and its API won't be able to run ...

Instead of `License submittal feature` it should be `some features of SPDX Online Tools` because credentials are also used by the XML Editor.

techytushar commented on README.md:
```diff
@@ -85,13 +82,21 @@ def getGithubSecret():
-* GHSECRETXXX is the Client ID for the Github Oauth Apps
+* GHSECRETXXX is the Client secret for the Github Oauth Apps
 * DJANGOSECRETXXX is the Django secret
+* OauthAppIDXXX is the client ID of the django oauth toolkit app
```
We can add a note here or add a link to the below section where it is described how to setup the django oauth app, otherwise users may get confused.

techytushar commented on README.md:
```diff
@@ -85,13 +82,21 @@ def getGithubSecret():
+* GHSECRETXXX is the Client secret for the Github Oauth Apps
 * DJANGOSECRETXXX is the Django secret
+* OauthAppIDXXX is the client ID of the django oauth toolkit app
+* OauthAppSecretXXX is the client secret of the django oauth toolkit app
```
Same here

Ugtan commented:
@goneall @techytushar I have updated the documentation with the suggested changes. Please have a look.
